File: src/camera_node.cpp
#include "aios/camera_node.h"
#include "aios/event_bus.h"

#include <chrono>
#include <cstdio>

namespace aios {

CameraNode::CameraNode(const std::string& path, int owner_uid, int permissions)
    : VfsNode(VfsNodeType::CAMERA, path, owner_uid, permissions) {}

std::string CameraNode::read() const {
    std::lock_guard<std::mutex> lock(mutex_);
    capture_count_++;

    auto now = std::chrono::system_clock::now();
    auto epoch = std::chrono::duration_cast<std::chrono::seconds>(
        now.time_since_epoch()).count();

    std::string mock_frame =
        "{\n"
        "  \"device\": \"" + path_ + "\",\n"
        "  \"capture_id\": " + std::to_string(capture_count_) + ",\n"
        "  \"timestamp\": " + std::to_string(epoch) + ",\n"
        "  \"resolution\": \"1920x1080\",\n"
        "  \"format\": \"JPEG\",\n"
        "  \"mode\": \"mock_capture\",\n"
        "  \"description\": \"A mock camera frame showing a desk with a monitor displaying AIOS kernel logs, a keyboard, and a cup of coffee.\",\n"
        "  \"base64_preview\": \"iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==\"\n"
        "}\n";

    std::printf("[CameraNode] READ %s | capture #%d | ts=%lld\n",
                path_.c_str(), capture_count_, static_cast<long long>(epoch));

    EventBus::instance().publish(EventType::VFS_WRITE, "CameraNode",
        path_ + " | capture #" + std::to_string(capture_count_));

    return mock_frame;
}

bool CameraNode::write(const std::string& data) {
    std::printf("[CameraNode] WRITE %s | ignored (camera is read-only device) | %zu bytes\n",
                path_.c_str(), data.size());
    return false;
}

} // namespace aios

File: aios-java/src/main/java/com/ouisani/aios/vfs/CameraNode.java
package com.ouisani.aios.vfs;

import com.ouisani.aios.core.VfsNode;

public non-sealed class CameraNode implements VfsNode {

    private final String path;
    private int ownerUid;
    private int permissions;

    public CameraNode(String path) {
        this(path, 0, 0444);
    }

    public CameraNode(String path, int ownerUid, int permissions) {
        this.path = path;
        this.ownerUid = ownerUid;
        this.permissions = permissions;
    }

    @Override
    public VfsNodeType nodeType() {
        return VfsNodeType.CAMERA;
    }

    @Override
    public String path() {
        return path;
    }

    @Override
    public int ownerUid() {
        return ownerUid;
    }

    @Override
    public void setOwnerUid(int uid) {
        this.ownerUid = uid;
    }

    @Override
    public int permissions() {
        return permissions;
    }

    @Override
    public void setPermissions(int perm) {
        this.permissions = perm;
    }

    @Override
    public String read() {
        String json = "{\"type\":\"image_b64\", \"scene\": \"一名黑客正在盯着屏幕敲代码\", \"timestamp\": " + System.currentTimeMillis() + "}";
        log.debug("CameraNode.read: path={}, captured frame", path);
        return json;
    }

    @Override
    public boolean write(String data) {
        throw new UnsupportedOperationException("CameraNode is read-only: " + path);
    }
}
